ios/engine/KMEI/KeymanEngine/Classes/KeymanTextField.swift

```diff
@@ -0,0 +1,355 @@
+  public override var inputView: UIView? {
+    get {
+      Manager.shared.webDelegate = self
+      return Manager.shared.inputView
+    }
+
+    set(inputView) {
+      super.inputView = inputView
+    }
+  }
+
+  // Prevent the delegate being set to anything but the delegateProxy
+  public override weak var delegate: UITextFieldDelegate? {
+    get {
+      return super.delegate
+    }
+
+    set(delegate) {
+      // Allow clearing of delegate
+      guard let delegate = delegate else {
+        super.delegate = nil
+        return
+      }
+
+      if delegate !== delegateProxy {
+        Manager.shared.kmLog(
+          "Trying to set KeymanTextField's delegate directly. Use setKeymanDelegate() instead.",
+          checkDebugPrinting: true)
+      }
+      super.delegate = delegateProxy
+    }
+  }
+
+  // MARK: - Public Methods
+
+  // Use this KeymanTextFieldDelegate instead of the normal UITextFieldDelegate.
+  // All of the normal UITextFieldDelegate methods are supported.
+  public func setKeymanDelegate(_ keymanDelegate: KeymanTextFieldDelegate?) {
+    delegateProxy.keymanDelegate = keymanDelegate
+    Manager.shared.kmLog(
+      "KeymanTextField: \(self.debugDescription) keymanDelegate set to: \(keymanDelegate.debugDescription)",
+      checkDebugPrinting: true)
+  }
+
+  // Dismisses the keyboard if this textview is the first responder.
+  //   - Use this instead of [resignFirstResponder] as it also resigns the Keyman keyboard's responders.
+  @objc public func dismissKeyboard() {
+    Manager.shared.kmLog(
+      "KeymanTextField: \(self.debugDescription) Dismissing keyboard. Was first responder:\(isFirstResponder)",
+      checkDebugPrinting: true)
+    resignFirstResponder()
+    Manager.shared.inputView.endEditing(true)
+  }
+
+  public override var text: String! {
+    get {
+      return super.text ?? ""
+    }
+
+    set(text) {
+      if let text = text {
+        super.text = text
+      } else {
+        super.text = ""
+      }
+
+      Manager.shared.setText(self.text)
+      let textRange = selectedTextRange ?? UITextRange()
+      let newRange = NSRange(location: offset(from: beginningOfDocument, to: textRange.start),
+                             length: offset(from: textRange.start, to: textRange.end))
+      Manager.shared.setSelectionRange(newRange, manually: false)
+    }
+  }
+
+  // MARK: - KMWebViewDelegate
+  @objc public func updatedFragment(_ fragment: String) {
+    if fragment.contains("insertText") {
+      processInsertText(fragment)
+    } else if fragment.contains("hideKeyboard") {
+      dismissKeyboard()
+    } else if fragment.contains("menuKeyUp") {
+      if let viewController = viewController {
+        Manager.shared.showKeyboardPicker(in: viewController, shouldAddKeyboard: false)
+      } else {
+        Manager.shared.switchToNextKeyboard()
+      }
+    }
+  }
+
+  private func processInsertText(_ fragment: String) {
+    if Manager.shared.isSubKeysMenuVisible {
+      return
+    }
```

Comment:
You know, a lot of this [looks really familiar](https://github.com/keymanapp/keyman/pull/378/files#diff-0b5708f03e4687dc51761c5c725f7ea4R159)...  I do recognize that the `if dn <= 0` part and after are different, though.

Some refactoring into a utility helper method for 'fragment' processing might be nice.

Reply:
Working on abstracting KMW. That will result in the fragment parsing being removed from here.